### ISSUE TITLE
Audit failure during block reorganizations

### ISSUE THREAD

**Issue by dexX7** [zathras-crypto/omnicore#66]
Re: the Travis failures:

It looks like this is caused by the Bitcoin Core RPC tests, which involve reorganizations and network splits. In particular:

```
Auditor has detected processing of a block in a non-sequential order (201)
```

```
Auditor has received unexpected notification of completion of block 201
```

```
Auditor has detected inconsistencies when attempting to update the total amount of tokens for property 1
Type:increase  Total cached:3815712997063, Amount changed:2640094941521, Total cached after update:6455807938584, Total from state:3815712997063
```

This is probably where it happens, in this test at least: [mempool_resurrect_test.py#L67-L70](https://github.com/zathras-crypto/mastercore/blob/db0916bd13d7a7f932fecd1b41c3cc1c8c61264a/qa/rpc-tests/mempool_resurrect_test.py#L67-L70)

To reproduce, dump the information to the console, and run:

```
omnicore/qa/rpc-tests$ ./mempool_resurrect_test.py
```

... or alternatively, don't let the test remove the temporary data and specify a target datadir:

```
omnicore/qa/rpc-tests$ ./mempool_resurrect_test.py --nocleanup --tmpdir=./testoutput
```

Then, you'd find the logs and data in:

```
omnicore/qa/rpc-tests/testoutput/node0/regtest/mastercore.log
omnicore/qa/rpc-tests/testoutput/node1/regtest/mastercore.log
omnicore/qa/rpc-tests/testoutput/node2/regtest/mastercore.log
omnicore/qa/rpc-tests/testoutput/node3/regtest/mastercore.log
```

One quick note regarding those tests in general: usually, during the first run, a cache is created for the first 200 blocks. Not all tests use that cache, but many do. This cache dir should be in `omnicore/qa/rpc-tests/cache` and you can start with a completely clean state by removing that dir:

```
omnicore/qa/rpc-tests$ rm -rf cache/
```

There are probably other reorg tests in the same dir `omnicore/qa/rpc-tests/`, and Travis also runs additional tests of BitcoinJ, via the [BitcoindComparisionTool](https://github.com/TheBlueMatt/test-scripts/), but those Python tests are much easier to run. :)

**Comment by dexX7** [comment]
Ah, special note about `--nocleanup`: it currently does not shut down the nodes properly, so you have kill the open processes manually, or adopt the following commit: https://github.com/dexX7/bitcoin/commit/7d1bf7c598eaf8b676e9c9ef4e0616f181ed3c9b

**Comment by zathras-crypto** [comment]
Thanks for the info dude, I'm going to take a look at auditor reorg handling today...   Speaking of bitcoin reorg tests, do we have the same for the Omni side?

**Comment by zathras-crypto** [comment]
This was really helpful :)  Thanks.

Was able to test 404b8f71dcd1245e9c4746e92512022060ac9683 successfully following the above.

```
zathras@coredev01:~/github/build/omnicore/qa/rpc-tests$ ./mempool_resurrect_test.py 
Initializing test directory /tmp/test3bZ5so
2015-04-22 23:28:53 Initializing Omni Core v0.0.9.2-dev [regtest]
2015-04-22 23:28:53 Loading trades database: OK
2015-04-22 23:28:53 Loading send-to-owners database: OK
2015-04-22 23:28:53 Loading transactions database: OK
2015-04-22 23:28:53 Loading smart property database: OK
2015-04-22 23:28:53 ##################################################################################
2015-04-22 23:28:53 Auditor initializing...
2015-04-22 23:28:53 Auditor initialized
2015-04-22 23:28:53 Exodus balance: 11756.18055542 MSC
2015-04-22 23:28:53 Omni Core initialization completed
2015-04-22 23:28:54 isMPinBlockRange(201, 202); n_found= 0
2015-04-22 23:28:54 deleteAboveBlock(200); tradedb n_found= 0
2015-04-22 23:28:54 deleteAboveBlock(200); stodb n_found= 0
2015-04-22 23:28:54 Auditor was notified that the state has been rolled back to block 200 due to a reorg/orphan. The auditor will now restart
2015-04-22 23:28:54 ##################################################################################
2015-04-22 23:28:54 Auditor initializing...
2015-04-22 23:28:54 Auditor initialized
Cleaning up
2015-04-22 23:28:54 Omni Core shutdown completed
Tests successful
```

**Comment by dexX7** [comment]
> Speaking of bitcoin reorg tests, do we have the same for the Omni side?

Now I can say: yes! ... I created some tests at least. :)

But there were complications: even though Bitcoin Core 0.10 has hidden RPC commands to invalidate and roll back blocks, there is no way (that I know of), to clear the memory pool, and after invalidating blocks, the transactions are added back to the pool. This limited the testing to some degree, because once a new block was mined, the transactions were picked up again, while they shouldn't.

I might think about another workaround, but given that the OmniJ CI server currently doens't use Omni Core 0.0.10, I just added a new RPC command:
- https://github.com/dexX7/bitcoin/compare/e687023daf4c4720d6804d7dcc51b8df2bc55684...dexX7:oc-0.10-rpc-clearmempool

... and a few tests for simple send, send to owner and property creation transactions:
- https://github.com/dexX7/bitcoin-spock/compare/b1927bea3272b8d18bcbd61a4d296839c9247486...dexX7:rpc-reorg-tests

I'm not sure, if this is of any use for you, but I'm relieved about the confirmation that Omni Core really passes the tests.

**Comment by dexX7** [comment]
The auditor works fine with the tests.

As a side note, because this is basically the main reason to clear the memory pool: after invalidating a block, the Omni Core state doesn't seem to be changed, but the full effects are only visible, once a new block is mined. In practise, there is probably no difference: another chain "wins", and this sort of implies "mining a new block".

**Comment by zathras-crypto** [comment]
> I'm not sure, if this is of any use for you

Are you kidding?  :)  Hugely useful mate :)  I didn't do the reorg stuff so I'd always wondered how solid it was.  The guy who wrote it (Bart) is very good (well above my grade haha), but I'd always wondered if we were properly tested in that area.  I'm VERY glad to hear you spin up some tests and that they pass.

Incidentally, just for sanity a while back I put in some quick detection that we're not leaving stuff behind in a reorg - does `mastercore.log` contain any lines with the text `LEVELDB TX OVERWRITE DETECTION - txid1234abcd` when you run this test?  I'd expect so as I haven't pushed https://github.com/zathras-crypto/mastercore/commit/4ef9c6de3891b81c97d5e0c5700f0fd15ddd5836 to main yet...

**Comment by zathras-crypto** [comment]
Also thought I'd ask - did you know the bitcoindcomparisontool much?  I'm trying to figure out what scenario is being run here that (seems to) cause `mastercore_handler_block_begin` to be called twice for the same block...

```
03:46:37 1 BitcoindComparisonTool.main: Testing block b7 6f31d24d520b6aee6c8166bfe9a83dc07aa18c61b0a5ac281b21452deb7127ae
03:46:37 1 AbstractBlockChain.connectBlock: Block forks the chain at height 103/block 042bd62a8367dcdc22b37c62bb041effee3d3c9a7848e3a2399827c52032f4ff, but it did not cause a reorganize:
6f31d24d520b6aee6c8166bfe9a83dc07aa18c61b0a5ac281b21452deb7127ae
03:46:37 1 BitcoindComparisonTool.main: Sent inv with block 6f31d24d520b6aee6c8166bfe9a83dc07aa18c61b0a5ac281b21452deb7127ae
03:46:37 15 BitcoindComparisonTool$1.onPreMessageReceived: Found header 042bd62a8367dcdc22b37c62bb041effee3d3c9a7848e3a2399827c52032f4ff
03:46:37 15 BitcoindComparisonTool$1.onPreMessageReceived: Sending header (042bd62a8367dcdc22b37c62bb041effee3d3c9a7848e3a2399827c52032f4ff) -> 6f31d24d520b6aee6c8166bfe9a83dc07aa18c61b0a5ac281b21452deb7127ae
03:46:37 15 BitcoindComparisonTool$1.onPreMessageReceived: Requested 6f31d24d520b6aee6c8166bfe9a83dc07aa18c61b0a5ac281b21452deb7127ae
03:46:37 15 BitcoindComparisonTool$1.onPreMessageReceived: Got empty header message from bitcoind
03:46:37 1 BitcoindComparisonTool.main: Block "b7" completed processing
03:46:37 1 BitcoindComparisonTool.main: Testing block b8 61dc8f54c5cd602b1935b6ca63aeeb7b2caf63b7eb43d39fde377d0fc23a5245
03:46:37 1 AbstractBlockChain.connectBlock: Block is causing a re-organize
03:46:37 1 AbstractBlockChain.handleNewBestChain: Re-organize after split at height 103
03:46:37 1 AbstractBlockChain.handleNewBestChain: Old chain head: 3780cf49ca30750a90b4fa5a590dc3dd44f013371ad7623c955f448f89f884a6
03:46:37 1 AbstractBlockChain.handleNewBestChain: New chain head: 61dc8f54c5cd602b1935b6ca63aeeb7b2caf63b7eb43d39fde377d0fc23a5245
03:46:37 1 AbstractBlockChain.handleNewBestChain: Split at block: 042bd62a8367dcdc22b37c62bb041effee3d3c9a7848e3a2399827c52032f4ff
03:46:38 1 BitcoindComparisonTool.main: Sent inv with block 61dc8f54c5cd602b1935b6ca63aeeb7b2caf63b7eb43d39fde377d0fc23a5245
03:46:38 15 BitcoindComparisonTool$1.onPreMessageReceived: Found header 042bd62a8367dcdc22b37c62bb041effee3d3c9a7848e3a2399827c52032f4ff
03:46:38 15 BitcoindComparisonTool$1.onPreMessageReceived: Sending header (042bd62a8367dcdc22b37c62bb041effee3d3c9a7848e3a2399827c52032f4ff) -> 6f31d24d520b6aee6c8166bfe9a83dc07aa18c61b0a5ac281b21452deb7127ae
03:46:38 15 BitcoindComparisonTool$1.onPreMessageReceived: Sending header (6f31d24d520b6aee6c8166bfe9a83dc07aa18c61b0a5ac281b21452deb7127ae) -> 61dc8f54c5cd602b1935b6ca63aeeb7b2caf63b7eb43d39fde377d0fc23a5245
03:46:38 15 BitcoindComparisonTool$1.onPreMessageReceived: Requested 61dc8f54c5cd602b1935b6ca63aeeb7b2caf63b7eb43d39fde377d0fc23a5245
2015-04-23 03:46:38 isMPinBlockRange(104, 104); n_found= 0
2015-04-23 03:46:38 deleteAboveBlock(103); tradedb n_found= 0
2015-04-23 03:46:38 deleteAboveBlock(103); stodb n_found= 0
2015-04-23 03:46:38 Auditor was notified that the state has been rolled back to block 103 due to a reorg/orphan. The auditor will now restart
2015-04-23 03:46:38 ##################################################################################
2015-04-23 03:46:38 Auditor initializing...
2015-04-23 03:46:38 Auditor initialized
2015-04-23 03:46:38 Auditor was notified block 104 has begun processing
2015-04-23 03:46:38 Auditor was notified block 104 has begun processing
2015-04-23 03:46:38 Auditor has detected processing of a block in a non-sequential order (104)

Error: Shutting down due to audit failure.  Please check mastercore.log for details
```

Full build log https://s3.amazonaws.com/archive.travis-ci.org/jobs/59660504/log.txt

**Comment by dexX7** [comment]
Very strange, I actually wrote an answer here, but for some reason the comment is not here.. probably closed the tab.. :S

> Are you kidding? :)

I was wondering, if this is useful for _you_, and if you were able to use it. :) The reorg tests were merged by @msgilligan into OmniJ, so here is another step-by-step instruction (starting from scratch, and you probably can save a few steps):
- https://gist.github.com/dexX7/1bacb0b4a0aff2ff432c

> LEVELDB TX OVERWRITE DETECTION

Hmm.. after running the tests, I actually don't see it, but I saw it before: when the initial parsing is stopped, and restarted, then it starts from the very beginning, which appears to trigger the overwrite detection.

> The guy who wrote it (Bart) is very good

Yes, I totally agree! His persistence is also very clever, but what I don't like is using plain (text) files, and I'm wondering for quite a while, why we use files _and_ LevelDB. You don't no more about this, do you? :)

> bitcoindcomparisontool

It's sort of a black box to me. As far as I know, it's somehow initially based on [BitcoinJ's tests](https://github.com/bitcoinj/bitcoinj/blob/master/core/src/test/java/org/bitcoinj/core/BitcoindComparisonTool.java), but it's also [directly available](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/depends/packages/native_comparisontool.mk) as Bitcoin Core dependency package (this is used by Travis), and it can be build into Bitcoin/Omni Core via [the configuration](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/configure.ac#L101-L109).

But for what it's worth, you can run it locally, instead of using Travis, to check the results:
- https://gist.github.com/dexX7/c2c0c1ffdc4b2fa942b0

**Comment by dexX7** [comment]
It looks like the error might be related to the order of calls:
- ConnectTip
- mastercore_handler_block_begin
- Failure to connect the new block

This raises the general question, whether the Omni Core handlers are properly placed. In this case, as first instinct, I'd say: we actually only want to trigger a block handler call, if the block is going to be accepted.

With the following modification, the BitcoindComparisonTool test seems to pass, and the OmniJ tests as well:

``` diff
diff --git a/src/main.cpp b/src/main.cpp
index e89a9c5..cc0f87c 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -1997,16 +1997,6 @@ static int64_t nTimePostConnect = 0;
  */
 bool static ConnectTip(CValidationState &state, CBlockIndex *pindexNew, CBlock *pblock) {
     assert(pindexNew->pprev == chainActive.Tip());
-
-    //! Omni Core: transaction position within the block
-    unsigned int nTxIdx = 0;
-    //! Omni Core: number of meta transactions found
-    unsigned int nNumMetaTxs = 0;
-
-    //! Omni Core: begin block connect notification
-    LogPrint("handler", "Omni Core handler: block connect begin [height: %d]\n", GetHeight());
-    mastercore_handler_block_begin(GetHeight(), pindexNew);
-
     mempool.check(pcoinsTip);
     // Read block from disk.
     int64_t nTime1 = GetTimeMicros();
@@ -2042,6 +2032,15 @@ bool static ConnectTip(CValidationState &state, CBlockIndex *pindexNew, CBlock *
         return false;
     int64_t nTime5 = GetTimeMicros(); nTimeChainState += nTime5 - nTime4;
     LogPrint("bench", "  - Writing chainstate: %.2fms [%.2fs]\n", (nTime5 - nTime4) * 0.001, nTimeChainState * 0.000001);
+
+    //! Omni Core: transaction position within the block
+    unsigned int nTxIdx = 0;
+    //! Omni Core: number of meta transactions found
+    unsigned int nNumMetaTxs = 0;
+    //! Omni Core: begin block connect notification
+    LogPrint("handler", "Omni Core handler: block connect begin [height: %d]\n", GetHeight());
+    mastercore_handler_block_begin(GetHeight(), pindexNew);
+
     // Remove conflicting transactions from the mempool.
```

Edit: ideally, we'd connect to the core signals directly: [main.cpp#L161-L176](https://github.com/OmniLayer/omnicore/blob/d6a48bf907518df8e37b2f64a32295406c60cf59/src/main.cpp#L161-L176)

The handlers are currently before, after, and very close to the `SyncTransaction` (via `SyncWithWallets`) signal.

**Comment by msgilligan** [comment]
I think it's time to migrate the Jenkins CI server to an 0.10.x blockchain: https://github.com/OmniLayer/OmniJ/issues/73

**Comment by zathras-crypto** [comment]
> With the following modification

Yep I see I see, so currently due to the location we call the handler (at the very start) it is still called even if ConnectTip is aborted for any reason (invalid block, read/write failure etc).  Moving it makes the most sense agreed.

https://github.com/zathras-crypto/mastercore/commit/b802850bdf68b8eba51405eee5ff9c0324d627d1

EDIT: forgot the magic word hehe - thanks :)

> Edit: ideally, we'd connect to the core signals directly

I'm not so sure about this - what are the benefits of switching to a message queue model rather than the current (directly calling functions) model?  Eg `ConnectTip` isn't called by a signal, it's called directly.  I just worry that moving to a message queue for block and transaction handling is going to open a potential can of worms with regard to ordering of actions.  I'm not completely opposed, but I'd like to understand a bit more what we gain :)

**Comment by zathras-crypto** [comment]
> Hmm.. after running the tests, I actually don't see it

That's interesting - do the tests actually orphan a block _containing a Omni transaction_?  If so that transaction should be written to levelDB, `deleteAboveBlock` won't catch it (without the fix I mentioned) and thus when the transaction is reprocessed in a new block it should be an overwrite and thus log that message.

> but I saw it before: when the initial parsing is stopped, and restarted, then it starts from the very beginning, which appears to trigger the overwrite detection.

Hmm, that's not cool - that indicates the state isn't being flushed (eg `clear_all_state`) if parsing is stopped and restarted from the beginning.

**Comment by zathras-crypto** [comment]
> Yes, I totally agree! His persistence is also very clever, but what I don't like is using plain (text) files, and I'm wondering for quite a while, why we use files and LevelDB. You don't no more about this, do you? :)

Well, I'd assume this was more evolution than revolution.  The very early days we actually had a flat file based on a specific block that we would load in balances from (before we had crowdsale code and thus weren't able to parse from the beginning) - it's stretching my memory a bit haha but I think it was simply that persistence evolved to where it is today through a couple iterations and it all started from those rather simple flat files.

**Comment by dexX7** [comment]
> I'm not so sure about this - what are the benefits of switching to a message queue model rather than the current (directly calling functions) model?

Let me rephrase: the handlers mostly operate on a "per block" basis, rather than a "per transaction" basis. This isn't bad, since the whole system is sort of "block oriented", but I think in the long run (and I mentioned it as some potential improvement for the future), we might refine the handling of transactions, in particular of unconfirmed ones, and could detect in greater detail, when something happens, whether it's the confirmation of a transaction/block or dropping of transactions for some reason. I haven't really looked at the signals yet though.

> That's interesting - do the tests actually orphan a block containing a Omni transaction?
> Hmm, that's not cool - that indicates the state isn't being flushed (eg clear_all_state) if parsing is stopped and restarted.

Let me check this again. I'll report back in about an hour.

> Well, I'd assume this was more evolution than revolution ... I think it was simply that persistence evolved to where it is today.

Ah, this makes sense. I just saw the whole picture in retrospective, so to speak. :)

**Comment by dexX7** [comment]
> That's interesting - do the tests actually orphan a block containing a Omni transaction?

To clarify: the regtests I created roll back the state and check, whether the state was rolled back. The transactions are not added twice.

During the Bitcoin Core tests (e.g. BitcoindComparisonTool) there are no Omni transactions involved.

**Comment by zathras-crypto** [comment]
Yeah, sorry my bad.  Basically overwrite detection only applies to `p_txlistdb` but the deletions from that are handled by `isMPinBlockRange()`.

`deleteAboveBlock` is only used for the trade and STO databases which don't have overwrite detection.

TL:DR; the bug I was talking about that is fixed by https://github.com/zathras-crypto/mastercore/commit/4ef9c6de3891b81c97d5e0c5700f0fd15ddd5836 only applies to trade and STO databases so you won't see the overwrite notice in the logs.

Sorry for the wasted time mate.

**Comment by dexX7** [comment]
> Hmm, that's not cool - that indicates the state isn't being flushed (eg clear_all_state) if parsing is stopped and restarted.

I'm a bit surprised: I logged that `load_most_relevant_state()` fails, because `_my_sps` has no watermark, and `clear_all_state()` is called before the initial scanning, but nevertheless I see the `"LEVELDB TX OVERWRITE DETECTION"` entries. Where are those transactions coming from?

**Comment by dexX7** [comment]
Ah: `p_txlistdb` isn't cleared at all. Is this expected?

**Comment by zathras-crypto** [comment]
Hmm, neither are the trade or sto DB's.

>  Is this expected?

Nope.  Looking now...

**Comment by zathras-crypto** [comment]
Looks like

```
t_tradelistdb->clear();
s_stolistdb->clear();
p_txlistdb->clear();
```

needs to be added to `clear_all_state`

**Comment by dexX7** [comment]
There has to be a `clear()` at first, but I'll take a look at it.

Is it correct that `size_t nCacheSize`, `bool fMemory`, `bool fWipe` are unused in the constructors?

**Comment by zathras-crypto** [comment]
> Is it correct that size_t nCacheSize, bool fMemory, bool fWipe are unused in the constructors?

They're used when we init:

```
  t_tradelistdb = new CMPTradeList(GetDataDir() / "MP_tradelist", 1<<20, false, fReindex);
  s_stolistdb = new CMPSTOList(GetDataDir() / "MP_stolist", 1<<20, false, fReindex);
  p_txlistdb = new CMPTxList(GetDataDir() / "MP_txlist", 1<<20, false, fReindex);
```

EDIT:  Oh.  Hmm.  I thought `CMPSTOList`, `CMPTradeList` and `CMPTxList` were using `CLevelDBWrapper` but they're not...

So to answer your question, yes that appears to be correct - they don't appear to be used.

**Comment by dexX7** [comment]
I picked up the wrapper idea: https://github.com/OmniLayer/omnicore/compare/omnicore-0.0.10...dexX7:oc-0.10-clear-base-database

There is still some debug output, and I have one, or two other ideas, but it looks like this works, and the LevelDB state is cleared via `clear_all_state()`.

I'm actually a bit surprised that we got away with not properly resetting the state? Probably means the rest of the logic is so solid, so this wasn't even an issue.
